soc/arm/atmel_sam0/common/soc_samd2x.c

```diff
@@ -9,167 +9,474 @@
+				    n)
+
+static void soc_sam0_fdpll_init(void)
+{
+#ifdef DT_FIXED_CLOCK_FDPLL_CLOCK_FREQUENCY
+	u32_t reg;
+	u32_t in_frequency;
+	u32_t mul;
+
+#ifdef DT_FIXED_CLOCK_XOSC32K_CLOCK_FREQUENCY
+	if (__builtin_strcmp(DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER,
+			     DT_FIXED_CLOCK_XOSC32K_LABEL) == 0) {
+		in_frequency = DT_FIXED_CLOCK_XOSC32K_CLOCK_FREQUENCY;
+		reg = SYSCTRL_DPLLCTRLB_REFCLK_REF0;
+	} else
+#endif
+#ifdef DT_FIXED_CLOCK_XOSC_CLOCK_FREQUENCY
+	if (__builtin_strcmp(DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER,
+			     DT_FIXED_CLOCK_XOSC_LABEL) == 0) {
+		in_frequency = DT_FIXED_CLOCK_XOSC_CLOCK_FREQUENCY;
+		reg = SYSCTRL_DPLLCTRLB_REFCLK_REF1;
+	} else
 #endif
+	{
+		struct device *clk;
+
+		clk = device_get_binding(DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER);
+		__ASSERT(clk != NULL, "invalid FDPLL input clock");
+
+		clock_control_on(clk,
+				 (clock_control_subsys_t)SYSCTRL_GCLK_ID_FDPLL);
+		clock_control_get_rate(clk,
+				       (clock_control_subsys_t)SYSCTRL_GCLK_ID_FDPLL,
+				       &in_frequency);
+		reg = SYSCTRL_DPLLCTRLB_REFCLK_GCLK;
+	}
 
-	wait_gclk_synchronization();
+	__ASSERT(DT_FIXED_CLOCK_FDPLL_CLOCK_FREQUENCY >= in_frequency,
+		 "invalid FDPLL input frequency");
 
-	/* Route GCLK1 to multiplexer 1 */
-	GCLK->CLKCTRL.reg =
-	    GCLK_CLKCTRL_ID(0) | GCLK_CLKCTRL_GEN_GCLK1 | GCLK_CLKCTRL_CLKEN;
-	wait_gclk_synchronization();
+	/*
+	 * SAMD21 Manual, 36.13.7: Fractional Digital Phase Locked Loop
+	 * (FDPLL96M) Characteristics
+	 */
+	__ASSERT(in_frequency >= KHZ(32),
+		 "FDPLL input frequency minimum is 32 kHz");
+	__ASSERT(reg == SYSCTRL_DPLLCTRLB_REFCLK_REF1 ||
+		 in_frequency <= MHZ(2),
+		 "FDPLL input frequency maximum is 2 MHz");
+	BUILD_ASSERT_MSG(DT_FIXED_CLOCK_FDPLL_CLOCK_FREQUENCY >= MHZ(48),
+			 "FDPLL output frequency minimum is 48 MHz");
+	BUILD_ASSERT_MSG(DT_FIXED_CLOCK_FDPLL_CLOCK_FREQUENCY <= MHZ(96),
+			 "FDPLL output frequency maximum is 96 MHz");
+
+	/*
+	 * Pick the largest possible divisor for XOSC that gives
+	 * LDR less than its maximum.  This minimizes the impact of
+	 * the fractional part for the highest accuracy.  This also
+	 * ensures that the input frequency is less than the limit
+	 * (2MHz).
+	 */
+	if (reg == SYSCTRL_DPLLCTRLB_REFCLK_REF1) {
+		u32_t div = in_frequency;
 
-	SYSCTRL->DFLLCTRL.reg = SYSCTRL_DFLLCTRL_ENABLE;
-	while (!SYSCTRL->PCLKSR.bit.DFLLRDY) {
+		div /= (2 * DT_FIXED_CLOCK_FDPLL_CLOCK_FREQUENCY);
+		div *= 0x1000U;
+
+		/* Ensure it's over the minimum frequency */
+		div = MIN(div, in_frequency / (2U * KHZ(32)) - 1U);
+
+		if (div > 0x800U) {
+			div = 0x800U;
+		} else if (div < 1U) {
+			div = 1U;
+		}
+
+		in_frequency /= (2U * div);
+
+		reg |= SYSCTRL_DPLLCTRLB_DIV(div - 1U);
 	}
 
-	u32_t mul = (SOC_ATMEL_SAM0_MCK_FREQ_HZ +
-		     SOC_ATMEL_SAM0_GCLK1_FREQ_HZ / 2) /
-		    SOC_ATMEL_SAM0_GCLK1_FREQ_HZ;
+	/*
+	 * Don't need SYSCTRL_GCLK_ID_FDPLL32K because we're not using
+	 * the lock timeout (but it would be easy to put as CLOCK_1).
```

Comment:
I've only tested this on same54 so far, but there I need to source `FDPLL032K` if `FDPLL0` is sourced by a GCLK. instead of e.g. XOSC32K

Reply:
I'm not sure what I wrote there anymore - the effect I'm observing is that if GCLK1 is sourced by ULPOSC32K and FDPLL0 is sourced by GCLK1, it takes a couple of seconds for the FDPLL to LOCK & get READY.

Non-deterministic behavior lends itself to ill conclusions…

I think I clocked the system before from OSCULP32K without that a long startup delay, so I'll have to investigate a bit.

Reply:
Do you still get the same behavior if you do something XOSC32K->GCLK1->FDPLL0?  I'm wondering if it's just instability in the ULPOSC32K causing a slow lock.

If you do end up needing it, then you can do something like:
```dts
&fdpll {
	clocks = <&gclk1>, <&gclk2>;
};
```

Then use `DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER_0` and `DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER_1` for the reference and `FDPLL032K`. 

It works without one on SAMD21 and the manual says "The FDPLL96M requires a 32kHz clock from the GCLK when the FDPLL96M internal lock timer is used."  So I guess it doesn't explicitly say you can operate without one, but I certainly read it like that.  The SAMD5x manual says "When the optional lock timer timeout function is used, a 32KHz reference clock (GCLK_DPLL_32K) is also required." (28.5.3, p 773), so it'd be somewhat surprising if it needed it regardless.